src/libcore/task/wake.rs

```diff
@@ -63,21 +67,103 @@ pub struct RawWakerVTable {
     /// task. Calling `wake` on the resulting [`RawWaker`] should result in a wakeup
     /// of the same task that would have been awoken by the original [`RawWaker`].
-    pub clone: unsafe fn(*const ()) -> RawWaker,
+    clone: unsafe fn(*const ()) -> RawWaker,
 
     /// This function will be called when `wake` is called on the [`Waker`].
     /// It must wake up the task associated with this [`RawWaker`].
     ///
     /// The implemention of this function must not consume the provided data
     /// pointer.
-    pub wake: unsafe fn(*const ()),
+    wake: unsafe fn(*const ()),
+
+    /// This function gets called when a [`RawWaker`] gets dropped.
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    drop: unsafe fn(*const ()),
+}
 
+impl RawWakerVTable {
+    /// Creates a new `RawWakerVTable` from the provided `clone`, `wake`, and
+    /// `drop` functions.
+    ///
+    /// # `clone`
+    ///
+    /// This function will be called when the [`RawWaker`] gets cloned, e.g. when
+    /// the [`Waker`] in which the [`RawWaker`] is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a [`RawWaker`] and associated
+    /// task. Calling `wake` on the resulting [`RawWaker`] should result in a wakeup
+    /// of the same task that would have been awoken by the original [`RawWaker`].
+    ///
+    /// # `wake`
+    ///
+    /// This function will be called when `wake` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// The implemention of this function must not consume the provided data
+    /// pointer.
+    ///
+    /// # `drop`
+    ///
     /// This function gets called when a [`RawWaker`] gets dropped.
     ///
     /// The implementation of this function must make sure to release any
     /// resources that are associated with this instance of a [`RawWaker`] and
     /// associated task.
-    pub drop: unsafe fn(*const ()),
+    #[rustc_promotable]
+    #[unstable(feature = "futures_api",
+            reason = "futures in libcore are unstable",
+            issue = "50547")]
+    pub const fn new(
+        clone: unsafe fn(*const ()) -> RawWaker,
+        wake: unsafe fn(*const ()),
+        drop: unsafe fn(*const ()),
+    ) -> Self {
+        Self {
+            clone,
+            wake,
+            drop,
+        }
+    }
+}
+
+/// The `Context` of an asynchronous task.
+///
+/// Currently, `Context` only serves to provide access to a `&Waker`
+/// which can be used to wake the current task.
+pub struct Context<'a> {
+    waker: &'a Waker,
+    // Ensure we future-proof against variance changes by including
+    // an `&mut` reference with the provided lifetime.
+    _marker: PhantomData<&'a mut ()>,
```

Comment:
`&` and `&mut` have the same variance wrt. their *lifetime*, so this changes nothing.

If you want `Context` to be invariant wrt. `'a`, you need something like `fn(&'a mut ()) -> &'a mut ()`.

(We really should have "variance markers" so that you can write `_marker: InvariantLifetime<'a>`...)

Reply:
I prefer `&'a mut &'a ()`, as its gets the invariance from mutation, not from composing covariance and variance.
But also agree on the marker.

Reply:
> I prefer &'a mut &'a (), as its gets the invariance from mutation, not from composing covariance and variance.

I know... but I find that much harder to remember that than for function types.

Also, mutation *is* combining covariance and contravariance: references are covariant in the type used for reading from them and contravariant in the type used for writing to them. It's the same deal as the function type. We just don't have notation to properly distinguish the two aspects of the type in a mutable reference.

Reply:
Part of it is that I saw someone do this recently:
```rust
type Invariant<'a> = PhantomData<fn(&'a ()) -> &'a ()>;
```
And, that's not actually invariant.

The argument and return types have opposing variances, but because the type is *structural*, not *nominal*, the behavior can be confusing (i.e. `Invariant<'a>` is a subtype/supertype of `PhantomData<fn(&'b ()) -> &'c ()>` given the right relationships - either `a: 'b, 'c: 'a` or `b: 'a, 'a: 'c`).

> I know... but I find that much harder to remember that than for function types.

What I remember is "mutable is invariant" so "wrap in `&mut`" (or `Cell`, if you *also* want `!Sync`) but that takes a type which means for `'a` you need `&'a ()`, resulting in `&mut &'a ()` / `Cell<&'a ()>`.

`&mut &'a ()` also needs a lifetime for the `&mut` and `&'a mut &'a ()` is the only easy choice (`'static` won't work unless `'a = 'static`, and you don't want to use more lifetime names).

You can replace `&'a ()` with `&'a T` if you also have a type parameter you want to be invariant wrt, I suppose.

Reply:
Well here you go: https://internals.rust-lang.org/t/pre-rfc-marker-types-for-making-parameters-invariant/9607

Reply:
@RalfJung I'm not trying to make it invariant, I'm just trying to make references to it (`&'a mut Contex<'b>`) have the same variance as if `Context<'b>` contained an `&'b mut SomeThing`.

Reply:
So that in the future if we add a `&'b mut SomeThing` to `Context<'b>` it will be a backwards-compatible addition.

Reply:
But you already have a reference. I don't understand.

The variance of a lifetime in a reference is the same type for mutable and shared references.

Reply:
Great, then this whole thing is a no-op and doesn't matter. I was under the impression that `&'a mut &'a ()` had different variance from `&'a mut &'a mut ()`. In any case, the code that I wrote is obviously correct, at best helpful, and at worst a no-op so i'd prefer to just leave it.

Reply:
Wait, that's totally different from what you said earlier. Now you *are* talking about nested references. All I said is that `&'a ()` and `'a mut ()` have the same variance in `'a`. But that does *not* mean that `&'a &'a ()` and `&'a mut &'a mut ()` have the same variance in `'a`! They don't.

Currently, the `PhantomData` is redundant and `'a` is covariant. That means that, e.g., `Context<'static>` implicitly turns into `Context<'a>`. If you don't want that, you need to make the lifetime parameter invariant.

`&'a mut T` anyway is invariant in `T`, so when you are considering `&'a mut Context<'b>`, that expression is invariant in `'b` no matter what `Context` does.